streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -244,18 +271,112 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
+                    return moveablePredicate.canMoveStandbyTask(sourceState, destinationState, task.id(), kafkaStreamsAssignments);
+                })
+                .map(AssignedTask::id)
+                .sorted()
+                .collect(Collectors.toList());
+        };
+
+        final long startTime = System.currentTimeMillis();
+        boolean taskMoved = true;
+        int round = 0;
+        final RackAwareGraphConstructor<KafkaStreamsAssignment> graphConstructor = RackAwareGraphConstructorFactory.create(
+            applicationState.assignmentConfigs().rackAwareAssignmentStrategy(), taskIds);
+        while (taskMoved && round < STANDBY_OPTIMIZER_MAX_ITERATION) {
+            taskMoved = false;
+            round++;
+            for (int i = 0; i < kafkaStreamsAssignments.size(); i++) {
+                final UUID clientId1 = clientIds.get(i);
+                final KafkaStreamsAssignment clientState1 = kafkaStreamsAssignments.get(new ProcessId(clientId1));
+                for (int j = i + 1; j < kafkaStreamsAssignments.size(); j++) {
+                    final UUID clientId2 = clientIds.get(i);
+                    final KafkaStreamsAssignment clientState2 = kafkaStreamsAssignments.get(new ProcessId(clientId2));
+
+                    final String rack1 = clientRacks.get(clientState1.processId().id()).get();
+                    final String rack2 = clientRacks.get(clientState2.processId().id()).get();
+                    // Cross rack traffic can not be reduced if racks are the same
+                    if (rack1.equals(rack2)) {
+                        continue;
+                    }
+
+                    final List<TaskId> movable1 = getMovableTasks.apply(clientState1, clientState2);
+                    final List<TaskId> movable2 = getMovableTasks.apply(clientState2, clientState1);
+
+                    // There's no needed to optimize if one is empty because the optimization
+                    // can only swap tasks to keep the client's load balanced
+                    if (movable1.isEmpty() || movable2.isEmpty()) {
+                        continue;
+                    }
+
+                    final List<TaskId> taskIdList = Stream.concat(movable1.stream(), movable2.stream())
+                        .sorted()
+                        .collect(Collectors.toList());
+                    final List<UUID> clients = Stream.of(clientId1, clientId2).sorted().collect(Collectors.toList());
+
+                    final AssignmentGraph assignmentGraph = buildTaskGraph(
+                        assignmentsByUuid,
+                        clientRacks,
+                        taskIdList,
+                        clients,
+                        topicPartitionsByTaskId,
+                        crossRackTrafficCost,
+                        nonOverlapCost,
+                        false,
+                        false,
```

Comment:
followup PR: should be true, true I think?

Reply:
you're right, good catch!